p2p/bad_peer_test.go

```diff
@@ -0,0 +1,154 @@
+package p2p_test
+
+import (
+	"context"
+	"encoding/binary"
+	"path/filepath"
+	"testing"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	varint "github.com/multiformats/go-varint"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/cmd/node"
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/config"
+	"github.com/spacemeshos/go-spacemesh/events"
+	"github.com/spacemeshos/go-spacemesh/log"
+	ps "github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+)
+
+func TestPeerDisconnectForMessageResultValidationReject(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Make 2 node instances
+	conf1 := config.DefaultTestConfig()
+	conf1.DataDirParent = t.TempDir()
+	conf1.FileLock = filepath.Join(conf1.DataDirParent, "LOCK")
+	conf1.P2P.Listen = "/ip4/127.0.0.1/tcp/0"
+	app1, err := NewApp(&conf1)
+	require.NoError(t, err)
+	conf2 := config.DefaultTestConfig()
+	conf2.DataDirParent = t.TempDir()
+	conf2.FileLock = filepath.Join(conf2.DataDirParent, "LOCK")
+	conf2.P2P.Listen = "/ip4/127.0.0.1/tcp/0"
+	app2, err := NewApp(&conf2)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		app1.Cleanup(ctx)
+		app2.Cleanup(ctx)
+	})
+	g := errgroup.Group{}
+	g.Go(func() error {
+		return app1.Start(ctx)
+	})
+	<-app1.Started()
```

Comment:
would be better if this couldn't get stuck for whole test timeout if app can't start

Reply:
it is also better to use mocknet from libp2p. tcp stack was very flaky on ci runners in the past

Reply:
Some of the configuration is done in node, so to test that we need to use node. I think it's nice also to know that the peer dropping is actually working in an approximately real system as opposed to a much more contained test that would involve the mocknet.

I'm interested to know more about the flaky tcp stack, I've never heard of a tcp stack being flaky?

Reply:
>it is environment issue, i am pretty sure that you will see it again with this test

Maybe this has been resolved? I've managed to run this test 23000 times on CI without a single failure.